bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java

```diff
@@ -330,52 +305,84 @@ private synchronized void stopService(SamsungTvService service) {
+            return;
+        }
+
+        logger.debug("remoteDeviceAdded: {}, {}", device.getType().getType(), device.getIdentity().getDescriptorURL());
+
+        createService(device);
+        putOnline();
 
+        if ("MediaRenderer".equals(device.getType().getType())) {
             /* Check if configuration should be updated */
             if (configuration.macAddress == null || configuration.macAddress.trim().isEmpty()) {
                 String macAddress = WakeOnLanUtility.getMACAddress(configuration.hostName);
                 if (macAddress != null) {
                     putConfig(SamsungTvConfiguration.MAC_ADDRESS, macAddress);
-                    logger.debug("thingDiscovered, macAddress: {}", macAddress);
+                    logger.debug("remoteDeviceAdded, macAddress: {}", macAddress);
                 }
             }
             if (SamsungTvConfiguration.PROTOCOL_NONE.equals(configuration.protocol)) {
                 Map<String, Object> properties = RemoteControllerService.discover(configuration.hostName);
                 for (Map.Entry<String, Object> property : properties.entrySet()) {
                     putConfig(property.getKey(), property.getValue());
-                    logger.debug("thingDiscovered, {}: {}", property.getKey(), property.getValue());
+                    logger.debug("remoteDeviceAdded, {}: {}", property.getKey(), property.getValue());
                 }
             }
-
-            /*
-             * SamsungTV discovery services creates thing UID from UPnP UDN.
-             * When thing is generated manually, thing UID may not match UPnP UDN, so store it for later use (e.g.
-             * thingRemoved).
-             */
-            upnpThingUID = result.getThingUID();
-            logger.debug("thingDiscovered, thingUID={}, discoveredUID={}", this.getThing().getUID(), upnpThingUID);
-            checkAndCreateServices();
+            checkCreateManualConnection();
         }
     }
 
     @Override
-    public void thingRemoved(DiscoveryService source, ThingUID thingUID) {
-        if (thingUID.equals(upnpThingUID)) {
-            logger.debug("Thing Removed: {}", thingUID);
+    public void remoteDeviceUpdated(@Nullable Registry registry, @Nullable RemoteDevice device) {
+    }
+
+    @Override
+    public void remoteDeviceRemoved(@Nullable Registry registry, @Nullable RemoteDevice device) {
+        if (configuration.hostName == null || device == null || device.getIdentity() == null
+                || device.getIdentity().getDescriptorURL() == null
+                || !configuration.hostName.equals(device.getIdentity().getDescriptorURL().getHost())
+                || device.getType() == null) {
+            return;
+        }
+
+        logger.debug("remoteDeviceRemoved: {}, {}", device.getType().getType(),
+                device.getIdentity().getDescriptorURL());
+
+        if ("MediaRenderer".equals(device.getType().getType())) {
             shutdown();
```

Comment:
Some of the Samsung TV models have a faulty UPnP implementation as they don't publish SSDP alive message as they have promised and jUPnP library might make a decision to remove the device. If device is a MediaRenderer device, shutdown now remove all samsung tv related UPnP services and they might not come back as remoteDeviceAdded is not called by the UPnP library for those removed services. Earlier checkAndCreateServices was called which created all related services.

Reply:
What's your suggestion on what to do here then? Instead of doing shutdown/offline call `checkCreateManualConnection`? Or simply ignore that call and assume that services are still available?

Why is it different to the previous `thingRemoved` being called by the UPnPDiscoveryService when the UPnP service disappears from JUPnP?

Reply:
When remoteDeviceAdded is called for MediaRenderer devices, binding could try to create all related services like it was done earlier by checkAndCreateServices.

Reply:
Why should it try to create services, if all are still there?

And please for me to understand: Why is it different to the previous thingRemoved being called by the UPnPDiscoveryService when the UPnP service disappears from JUPnP?

Reply:
ThingRemoved is not the issue, but the thingDiscovered. 

1. Samsung TV binding is up and running (MediaRenderer service and e.g. MainTV service are up)
2. MediaRenderer device disappears, so JUPnP library calls thingRemoved for MediaRenderer device.
3. During thingRemoved method, binding remove **all** services (MediaRenderer and MainTV) by shutdown method.
4. MediaRenderer device is discovered again and remoteDeviceAdded is called.
5. Binding creates MediaRenderer service.
6. MainTV service is missing

Reply:
I'm not using Samsung TV binding my self at the moment as I have separated VLAN's and Samsung TV doesn't allow communication from another network. Some users have been stated  that 2.4 version have been working very reliable.

https://github.com/openhab/openhab2-addons/blob/2.4.0/addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java

2.4 binding implemented both DiscoveryListener and RegistryListener.

thingDiscovered -> create all available services
thingRemoved -> shutdown all services
remoteDeviceAdded -> create discovered service

Reply:
> users have been stated that 2.4 version have been working very reliable

Yup, 2.4 RC1 is very stable.  I tried to work with folks on this earlier this year to get 2.5 version to work which introduces new Samsung TV models.   I even put up a bounty to fix this . . .

Best, Jay

Reply:
> ThingRemoved is not the issue, but the thingDiscovered.

Sorry, but I still do not understand. Do you refer to this PR or about an issue that already exists in the current 2.5/3.0 code base?

Note that this PR only tries to fix the issue that the binding currently causes deadlocks that affect other bindings and the core runtime as a whole.
If there are issues to solve with unreliable TV detection due to the handling of SSDP messages (or their non-existence due to buggy TVs), please let us discuss those in a different issue.

Reply:
>Do you refer to this PR

Yes

>Note that this PR only tries to fix the issue that the binding currently causes deadlocks that affect other bindings and the core runtime as a whole.

Yes I know, but this PR also change the behaviour of the binding, which I think might cause some issues.

Let's try one more time

On master, thingDiscovered method calls checkAndCreateServices (line 362), which will go throw all existing UPnP devices and create all necessary Samsung tv services.

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L333-L364

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L246-L254

All Samsung tv related services are removed in thingRemoved method when service is MediaRenderer device

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L367-L373

https://github.com/openhab/openhab-addons/blob/f9b763210d67f286186eeded6c69e1001ed28d2b/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java#L183-L189


In this PR, thingDiscovered and thingRemoved are removed and it fully relie to UPnP events (remoteDeviceAdded and remoteDeviceRemoved). When remoteDeviceRemoved is done for MediaRenderer device, it shutdown all existing services, but remoteDeviceAdded does only create service which was found, so following could happen

1. Samsung TV binding is up and running (MediaRenderer service and e.g. MainTV service are up)
2. MediaRenderer device disappears, so JUPnP library calls remoteDeviceRemoved for MediaRenderer device.
3. remoteDeviceRemoved method remove all services (MediaRenderer and MainTV) by shutdown method.
4. MediaRenderer device is discovered again and remoteDeviceAdded method is called.
5. Binding creates MediaRenderer service.
6. MainTV service is still missing, so only half of the binding features are working 

So, when MediaRenderer device is added, binding should check and create all Samsung TV related services (then it match equally the master branch functionality) or remoteDeviceAdded should not shutdown all services, but only service which was removed.

Reply:
> Let's try one more time

Thanks a lot for your patience with me. 🙏 

> On master, thingDiscovered method calls checkAndCreateServices (line 362), which will go throw all existing UPnP devices and create all necessary Samsung tv services.

So if I understand correctly, this is exactly the feature you are missing now with that PR, right? My confusion might come from the fact that we are discussing this in the `remoteDeviceRemoved` method.

> All Samsung tv related services are removed in thingRemoved method 

Yes, that's what I [referred to above](https://github.com/openhab/openhab-addons/pull/9525#discussion_r549727193) - it is the very same as what happens in the code of this PR, that's why I couldn't see any problem with the `remoteDeviceRemoved` method.

> So, when MediaRenderer device is added, binding should check and create all Samsung TV related services 

That should be easy to do in the `remoteDeviceAdded` method then. But it still is not clear to me why this solves the problem you mentioned initially: "they don't publish SSDP alive message". If there are no SSDP messages, we will only see a call to `remoteDeviceRemoved`, which falsely takes the Thing OFFLINE, but will never see a call to `remoteDeviceAdded`, which would bring it back ONLINE again. That's why I am confused on what should be done.

Reply:
I didn't say that TV don't publish SSDP alive messages at all, but not like they have promised (max-age for advertisement), so UPnP library think that service is offline and then suddenly TV send the advertisement. Behaviour seems to be also service specific and that's why there are many questions on the forum like yesterday day everything worked fine and today I can change the volume but not the channel. Which of course is weird from user perspective as they don't understand that TV contains multiple UPnP services and each service support different features.

Reply:
Few years ago, I tried to find solution for this and found following from Cling page, but I never found time to test it. It will impact to all UPnP devices and binding, but maybe something clever could implemented to OH side or JUPnP.

http://4thline.org/projects/cling/core/manual/cling-core-manual.xhtml#section.RemoteMaxAge

Anyhow, idea was to increase the max age or disable it and then manually but thing offline state if several consecutive queries fails (TV power is OFF).

Reply:
Ok, thanks again @paulianttila.
I have now reverted everything from #7499 that was not directly related to replacing thing discovery methods by upnp device discovery methods.
This leaves the code hopefully working exactly the same way as before, while still not causing deadlocks anymore caused by the discovery listener registration.